design/authorization-handler.md

```diff
@@ -0,0 +1,601 @@
+
+## Compliance processor
+
+There will be a chain of policies configured in the admin console for a specific Authorization  Handler. The ClientAuthorizationContext can be processed by this policy chain. Each AuthorizationDetailsEntry will be processed by each of the policy and policy will decide if:
+
+* AuthorizationDetails is required to be rendered on the consent screen
+* AuthorizationDetails is unsupported/invalid/unauthorized and whole client request should be rejected
+
+Policy can also “enrich” and/or change authorizationDetails entries as sometimes client sent “incomplete” authorization_details and it is up to the AS (Keycloak server) to figure correct values and then in later stages, display them to the user on consent screen or add them to the token. Some use-case described in the [RAR specification Section 7.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar-07#section-7.1).
+
+Policy is also responsible for processing the consent screen (The moment when a user submits the consent screen, which can happen multiple times during a single “authentication session” due the fact that more consent screens would be possible with this proposal).
+
+### Naming
+
+Maybe ClientScopesPolicy is ok as a name? IMO any name containing "Authorization" (EG. ClientAuthorizationContextPolicy, AuthorizationDetailsPolicy) can bring the confusion with the authorization services, which also use "Policies"... More wider question if ClientAuthorizationContext itself is good as a name?
+
+### Configuration
+
+Policies can be configured at the client scope (or ClientAuthorizationHandler) level. In addition to the Mappers and Scope tabs, there can be another tab like Policies. By default, each client scope can have only Persistent Consent Policy automatically set (See below for Persistent Consent Policy details).
+
+### When
+
+The policies are processed after the user is authenticated and required actions processed. Hence before the render of consent screen. So policies have access to existing authenticated users and current clients.
+In addition, policies are processed after each confirmation of the consent screen (When the user presses “Submit” on the consent screen). This is needed as consent screens in some cases can contain user data and require input from the user (For example there will be a list of accounts on consent screen and users will need to choose one of them).
+
+We can see if we need another way to process at the beginning of the request to AuthorizationEndpoint even before the user is authenticated (EG. For ability to reject request quickly)
+
+
+### Pseudo-code
+
+```java
+interface ClientScopePolicy {
+
+	PolicyVote applyPolicy(ClientScopePolicyContext ctx);
```

Comment:
IMO, a better design is to follow the same approach used in authentication so that specific methods are provided to infer information from the handler?

Reply:
I am not sure what exactly you mean by "authentication" ? Did not you mean Authorization services or something else? Can you mention any more concrete class/use-cases?

Reply:
I meant authentication as you did for `AuthenticationFlowContext`. There you have methods such as `success`, `failure`, etc.

Reply:
Ah, I see. This approach should work. I am not 100% sure if it is better to use "context" based mechanism like Authentication SPI uses or the mechanism based on something like "vote", which is used for example in Client Policies (See `ClientPolicyConditionProvider.applyPolicy`) . It seems both have some advantages.
@dgozalo WDYT?

Reply:
I tend to prefer relying on implementations to expose their behaviour and data given a set of methods instead of being given a context to work with.

Any change to the context can have a huge impact on these classes, while extensibility can certainly be easier when only dealing with the internals of each implementation, or even just creating new ones. So I believe the `AuthenticationFlowContext` would be preferred.

But then again, I'd love to define a common Policy Engine that would manage **all** the policies executed in Keycloak.

As I can see right now, there are at least three implementations doing that.